src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Services/DefaultRazorSemanticTokensInfoService.cs

```diff
@@ -132,153 +128,19 @@ internal class DefaultRazorSemanticTokensInfoService : RazorSemanticTokensInfoSe
                 return null;
             }
 
-            var semanticVersion = await GetDocumentSemanticVersionAsync(documentSnapshot);
+            var data = ConvertSemanticRangesToSemanticTokensData(combinedSemanticRanges, codeDocument);
+            var tokens = new SemanticTokens { Data = data.ToImmutableArray() };
 
-            if (newResultId is null)
-            {
-                // If there's no C# in the Razor doc, we won't have a resultId returned to us.
-                // Just use a GUID instead.
-                newResultId = Guid.NewGuid().ToString();
-            }
-
-            var razorSemanticTokens = ConvertSemanticRangesToSemanticTokens(
-                combinedSemanticRanges, codeDocument, newResultId, isFinalizedCSharp);
-            UpdateRazorDocCache(textDocumentIdentifier.Uri, semanticVersion, newResultId, razorSemanticTokens);
-
-            return new SemanticTokens { ResultId = razorSemanticTokens.ResultId, Data = razorSemanticTokens.Data.ToImmutableArray() };
-        }
+            var semanticTokensResponse = new SemanticTokensResponse(documentVersion, range, tokens);
 
-        public override async Task<SemanticTokensFullOrDelta?> GetSemanticTokensEditsAsync(
-            TextDocumentIdentifier textDocumentIdentifier,
-            string? previousResultId,
-            CancellationToken cancellationToken)
-        {
-            var documentPath = textDocumentIdentifier.Uri.GetAbsolutePath();
-            if (documentPath is null)
+            // Cache the result so we can potentially avoid recomputation next time around.
+            _cachedResponses.Add(semanticTokensResponse);
+            if (_cachedResponses.Count > MaxCacheSize)
             {
-                return null;
+                _cachedResponses.RemoveAt(0);
```

Comment:
This kind of compaction ensures that once the cap is reached each insertion guarantees a deletion. That's why `MemoryCache` [removes half the entries](https://github.com/dotnet/razor-tooling/blob/4dba549b6958682d8ea0bf33fe18ba285f19faa0/src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/MemoryCache.cs#L69-L72) on compaction.